### PR TITLE
[ALLUXIO-2835] Add option for behavior of loading metadata

### PR DESCRIPTION
**Pull request by gpang**
https://alluxio.atlassian.net/browse/ALLUXIO-2835

### CODE REVIEW

**Comment by gpang** [comment]
Jenkins, test this please

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 1]
```diff
@@ -176,7 +178,8 @@ public boolean exists(AlluxioURI path, ExistsOptions options)
     FileSystemMasterClient masterClient = mFileSystemContext.acquireMasterClient();
     try {
       // TODO(calvin): Make this more efficient
```
Comment (jsimsa) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: @calvinjia what are your thoughts on making this more efficient?

File: core/client/fs/src/main/java/alluxio/client/file/options/ExistsOptions.java [hunk 2]
```diff
@@ -34,21 +37,46 @@ public static ExistsOptions defaults() {
   }
 
   private ExistsOptions() {
-    // No options currently
+    mLoadMetadataType = LoadMetadataType.Once;
```
Comment (jsimsa) [inline core/client/fs/src/main/java/alluxio/client/file/options/ExistsOptions.java]: should this use `Configuration.getEnum(PropertyKey.USER_FILE_METADATA_LOAD_TYPE, LoadMetadataType.class)` as well?

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java (hunk 1 above)
Comment (calvinjia) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: The issue was with using exceptions for control flow caused by reusing the `getStatus` method for exists.

File: core/client/fs/src/test/java/alluxio/client/file/options/GetStatusOptionsTest.java [hunk 3]
```diff
@@ -12,6 +12,9 @@
 package alluxio.client.file.options;
```
Comment (jsimsa) [inline core/client/fs/src/test/java/alluxio/client/file/options/GetStatusOptionsTest.java]: should the ExistsOptionsTest be updated as well?

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 4]
```diff
@@ -732,10 +756,27 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
             && ((InodeDirectory) inode).isDirectChildrenLoaded()) {
           loadMetadataOptions.setLoadDirectChildren(false);
         }
+      } else {
+        if (listStatusOptions.getLoadMetadataType() == LoadMetadataType.Never || (
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: optional: this is duplicated in the code and could be factored out into a method

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java [hunk 5]
```diff
@@ -732,10 +756,27 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
...
+                .isAbsent(path))) {
+          throw new FileDoesNotExistException(
+              ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(inodePath.getUri()));
+        }
       }
 
       loadMetadataIfNotExistAndJournal(inodePath, loadMetadataOptions, journalContext);
-      mInodeTree.ensureFullInodePath(inodePath, InodeTree.LockMode.READ);
+
+      boolean exists = false;
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: optional: this is duplicated in the code and could be factored out into a method

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   * Creates a new instance of {@link UfsAbsentPathAsyncCache}.
+   *
+   * @param mountTable the mount table
+   */
+  public UfsAbsentPathAsyncCache(MountTable mountTable) {
+    super(mountTable);
+    mCurrentPaths = new ConcurrentHashSet<>(8, 0.95f, 8);
+
+    mPool = new ThreadPoolExecutor(UFS_ABSENT_PATH_CACHE_THREADS, UFS_ABSENT_PATH_CACHE_THREADS, 1,
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: can you use a named constant for `1` as well? thanks

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        TimeUnit.MINUTES, new LinkedBlockingQueue<Runnable>(),
+        ThreadFactoryUtils.build("UFS-Absent-Path-Cache-%d", true));
+  }
+
+  @Override
+  public void addAbsentPath(AlluxioURI path) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: optional: this could just be `add` or `addPath` I think since the class is already called `AbsentPathCache`

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class caches paths in the UFS which do not exist.
+ */
+@ThreadSafe
+public class UfsAbsentPathCache {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: I would prefer to make this an interface and have two implementations of it (sync and async).

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This class caches paths in the UFS which do not exist.
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: `Cache for recording information about paths that are not present in UFS.`

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 8 above)
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: and add a factory that takes a boolean that determines whether the cache should be synchronous or asynchronous

Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: also, if the synchronous cache is not used anywhere, we could just have one implemnetation

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+    String[] components = PathUtils.getPathComponents(ufsUri.getPath());
+    if (components.length == 0) {
+      return;
+    }
+
+    // create a ufs uri of the root of the mount point.
+    AlluxioURI uri = mMountTable.getMountInfo(resolution.getMountId()).getUfsUri();
+    int mountPointBase = uri.getDepth();
+
+    // Traverse through the ufs path components, staring from the root, to find the first
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: starting

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+  }
+
+  /**
+   * Adds the given absent path into the cache. This will sequentially walk down the path to find
+   * the first component which does not exist in the ufs.
+   *
+   * @param path the absent path to add to the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void addAbsentPath(AlluxioURI path) throws InvalidPathException {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: optional: this could just be `add` or `addPath` I think since the class is already called `AbsentPathCache`

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }
+
+  /**
+   * Removes an absent path from the cache.
+   *
+   * @param path the path to remove from the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public void removeAbsentPath(AlluxioURI path) throws InvalidPathException {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: optional: this could just be `remove` or `removePath` I think since the class is already called `AbsentPathCache`

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+  /**
+   * Returns true if the given path is absent, according to this cache. A path is absent if one of
+   * its ancestors is absent.
+   *
+   * @param path the path to check
+   * @return true if the path is absent according to the cache
+   * @throws InvalidPathException if the path is invalid
+   */
+  public boolean isAbsent(AlluxioURI path) throws InvalidPathException {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: `contains`?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    mMountTable = mountTable;
+    mCaches = new ConcurrentHashMapV8<>(8, 0.95f, 8);
+  }
+
+  /**
+   * Removes the cache for the given mount id.
+   *
+   * @param mountId the mount id
+   */
+  public void removeMountPoint(long mountId) {
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: why do we have `removeMountPoint` but not `addMountPoint`? I feel like we should have both or neither and I am leaning towards neither (by garbage collecting caches lazily during `add` or `remove`)

File: core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 15]
```diff
@@ -658,10 +663,16 @@ public void listStatusWithLoadMetadataNever() throws Exception {
...
     // getFileId should load metadata automatically.
     AlluxioURI uri = new AlluxioURI("/mnt/local/dir1");
-    List<FileInfo> fileInfoList = mFileSystemMaster.listStatus(uri,
-        ListStatusOptions.defaults().setLoadMetadataType(LoadMetadataType.Never));
-    Assert.assertEquals(0, fileInfoList.size());
-    Assert.assertEquals(4, mFileSystemMaster.getNumberOfPaths());
+    try {
+      mFileSystemMaster.listStatus(uri,
+          ListStatusOptions.defaults().setLoadMetadataType(LoadMetadataType.Never));
+      Assert.fail("listStatus() with load metadata NEVER should throw exception.");
```
Comment (jsimsa) [inline core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: nit: I would simply say "exception expected" to avoid hardcoding class and enum name into the error message

File: core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 16]
```diff
@@ -658,10 +663,16 @@ public void listStatusWithLoadMetadataNever() throws Exception {
...
+    } catch (FileDoesNotExistException e) {
+      // Expected case.
+    }
+
+    // 3 directories exist. No additional directories.
```
Comment (jsimsa) [inline core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: this comment is redundant

File: core/server/master/src/test/java/alluxio/master/file/meta/UfsAbsentPathCacheTest.java [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.File;
+import java.util.Collections;
+
+/**
+ * Unit tests for {@link UfsAbsentPathCache}.
+ */
+public class UfsAbsentPathCacheTest {
```
Comment (jsimsa) [inline core/server/master/src/test/java/alluxio/master/file/meta/UfsAbsentPathCacheTest.java]: so we have a unit test for the sync cache, but use the async cache in production?

File: tests/src/test/java/alluxio/master/file/LoadMetadataTest.java [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+          durationMs < SLEEP_MS / 2);
+    }
+
+    if (!expectExists && expectLoadFromUfs) {
+      // The metadata is loaded from Ufs, but the path does not exist, so it will be added to the
+      // absent cache. Wait until the path shows up in the absent cache.
+      final UfsAbsentPathCache cache = Whitebox.getInternalState(
+          mLocalAlluxioClusterResource.get().getLocalAlluxioMaster().getMasterProcess()
+              .getMaster(FileSystemMaster.class), "mUfsAbsentPathCache");
+      CommonUtils.waitFor("path (" + path + ") to be added to absent cache",
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/master/file/LoadMetadataTest.java]: use a timeout

File: tests/src/test/java/alluxio/master/file/LoadMetadataTest.java [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+import org.junit.Test;
+import org.powermock.reflect.Whitebox;
+
+import java.io.File;
+import java.io.FileWriter;
+
+/**
+ * Tests the loading of metadata and the available options.
+ */
+public class LoadMetadataTest {
```
Comment (jsimsa) [inline tests/src/test/java/alluxio/master/file/LoadMetadataTest.java]: extends `BaseIntegrationTest`

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+ * This class caches paths in the UFS which do not exist.
+ */
+@ThreadSafe
+public class UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsAbsentPathCache.class);
+
+  /** The mount table. */
+  protected final MountTable mMountTable;
+  /** Stores a cache for each mount point (the key is mount id). */
+  protected final ConcurrentHashMapV8<Long, ConcurrentSkipListSet<String>> mCaches;
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: it seems like this will grow without a bound ... I suggest we only use a single cache, leveraging Guava's cache implementation (https://github.com/google/guava/wiki/CachesExplained) ... entries in the cache are key by UFS URI and store mount point ID as value so that they can be validated / invalidated lazily at lookup (as opposed to invalidating all entries when a mount point is removed)

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    if (components.length == 0) {
+      return;
+    }
+
+    // create a ufs uri of the root of the mount point.
+    AlluxioURI uri = mMountTable.getMountInfo(resolution.getMountId()).getUfsUri();
+    int mountPointBase = uri.getDepth();
+
+    // Traverse through the ufs path components, staring from the root, to find the first
+    // non-existing ufs path.
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Is there an optimization we can do where we start at the path Alluxio no longer has a persisted directory? Might be premature optimization.

File: core/common/src/thrift/file_system_master.thrift [hunk 22]
```diff
@@ -57,6 +57,10 @@ struct ListStatusTOptions {
   2: optional LoadMetadataTType loadMetadataType
 }
 
+struct GetStatusTOptions {
+  1: optional LoadMetadataTType loadMetadataType
```
Comment (calvinjia) [inline core/common/src/thrift/file_system_master.thrift]: What's the default value for this?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 23]
```diff
@@ -467,6 +467,8 @@
       create(Name.USER_FILE_DELETE_UNCHECKED, false);
   public static final PropertyKey USER_FILE_MASTER_CLIENT_THREADS =
       create(Name.USER_FILE_MASTER_CLIENT_THREADS, 10);
+  public static final PropertyKey USER_FILE_METADATA_LOAD_TYPE =
+      create(Name.USER_FILE_METADATA_LOAD_TYPE, "Once");
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Could you document this property?

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 24]
```diff
@@ -118,7 +119,8 @@ public FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
     URIStatus status;
     try {
       masterClient.createFile(path, options);
-      status = masterClient.getStatus(path);
+      status = masterClient.getStatus(path, GetStatusOptions.defaults().setLoadMetadataType(
+          LoadMetadataType.Never));
```
Comment (jsimsa) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: it makes sense for both createFile and createDirectory to do that (and we should probably do it before 1.5)

Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: This is not part of this PR.

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 25]
```diff
@@ -176,7 +178,8 @@ public boolean exists(AlluxioURI path, ExistsOptions options)
     FileSystemMasterClient masterClient = mFileSystemContext.acquireMasterClient();
     try {
       // TODO(calvin): Make this more efficient
-      masterClient.getStatus(path);
+      masterClient.getStatus(path,
+          GetStatusOptions.defaults().setLoadMetadataType(options.getLoadMetadataType()));
```
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: Done

File: core/client/fs/src/main/java/alluxio/client/file/options/ExistsOptions.java (hunk 2 above)
Comment (gpang) [inline core/client/fs/src/main/java/alluxio/client/file/options/ExistsOptions.java]: Yes, done.

File: core/client/fs/src/test/java/alluxio/client/file/options/GetStatusOptionsTest.java (hunk 3 above)
Comment (gpang) [inline core/client/fs/src/test/java/alluxio/client/file/options/GetStatusOptionsTest.java]: Updated

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 23 above)
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Done

File: core/common/src/thrift/file_system_master.thrift (hunk 22 above)
Comment (gpang) [inline core/common/src/thrift/file_system_master.thrift]: The default is `Once`.

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 4 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java (hunk 6 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java (hunk 7 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: I updated the api to `process`, but I though just `add` was confusing, since it was adding an absent path. Anyways, everything is renamed and hopefully it is simpler to understand.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+        return;
+      }
+      try {
+        if (isAbsent(mPath)) {
+          // this path his already considered absent, so no need to add it again.
+          return;
+        }
+        UfsAbsentPathAsyncCache.super.addAbsentPath(mPath);
+      } catch (InvalidPathException e) {
+        // Ignore the exception
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Mount table access sometimes throws `InvalidPathException`. This logic is updated now, and the exception is logged.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 9 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 8 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Done. I created an interface, with a single async implementation.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 20 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: I changed this to a guava cache.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 14 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Done. Now the cache is represented differently.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 11 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: This has been renamed to `process`.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 10 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 21 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: I think that is possible, but since there is never a guarantee that alluxio is in-sync with ufs, we can just process normally. Since this will be done in a threadpool, it should not impact the critical path.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 12 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: Updated api to `process`.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java (hunk 13 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathCache.java]: I originally thought `contains` would have been confusing, since if this contains a path, it means it doesn't exist in ufs. Do you think it won't be confusing?

File: core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 15 above)
Comment (gpang) [inline core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: Done

File: core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java (hunk 16 above)
Comment (gpang) [inline core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: removed

File: core/server/master/src/test/java/alluxio/master/file/meta/UfsAbsentPathCacheTest.java (hunk 17 above)
Comment (gpang) [inline core/server/master/src/test/java/alluxio/master/file/meta/UfsAbsentPathCacheTest.java]: Originally, I only had the sync version, and later added the async version. Now, there is only a test for the async implementation (sync implementation is removed).

File: tests/src/test/java/alluxio/master/file/LoadMetadataTest.java (hunk 19 above)
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/LoadMetadataTest.java]: Done

File: tests/src/test/java/alluxio/master/file/LoadMetadataTest.java (hunk 18 above)
Comment (gpang) [inline tests/src/test/java/alluxio/master/file/LoadMetadataTest.java]: Done

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: this code still seem to be duplicated (see lines 681 and 759)

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java (hunk 24 above)
Comment (jsimsa) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: I agree

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java [hunk 27]
```diff
@@ -0,0 +1,231 @@
...
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * This is a {@link UfsAbsentPathCache} which implements asynchronous addition and removal to the
+ * cache, since the processing of the path may be slow.
+ */
+@ThreadSafe
+public final class UfsAbsentPathAsyncCache implements UfsAbsentPathCache {
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Should this be called `AsyncUfsAbsentPathCache`?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(UfsAbsentPathAsyncCache.class);
+  /** Number of threads for the async pool. */
+  private static final int NUM_THREADS = 50;
+  /** Number of seconds to keep threads alive. */
+  private static final int THREAD_KEEP_ALIVE_SECONDS = 60;
+  /** Number of paths to cache. */
+  private static final int MAX_PATHS = 100000;
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Should these values be configurable or is the impact minimal as long as the numbers are reasonable?

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java (hunk 27 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Renamed. I originally thought that the actual cache was async, so I named it that way. But, since it is the async implementation of an interface, this also works.

File: core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java (hunk 28 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/UfsAbsentPathAsyncCache.java]: Added a parameter.

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java [hunk 29]
```diff
@@ -0,0 +1,234 @@
...
+   * @param mountInfo the associated {@link MountInfo} for the Alluxio path
+   * @return if true, further traversal of the descendant paths should continue
+   */
+  private boolean checkPath(AlluxioURI alluxioUri, MountInfo mountInfo) {
+    ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    Lock writeLock = rwLock.writeLock();
+    Lock readLock = null;
+    try {
+      // Write lock this path, to only enable a single task per path
+      writeLock.lock();
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: at this point the lock is a local variable, so it will not achieve mutual exclusion

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java [hunk 30]
```diff
@@ -0,0 +1,234 @@
...
+        writeLock.unlock();
+        writeLock = null;
+        readLock = existingLock.readLock();
+        readLock.lock();
+      } else {
+        // This thread has the exclusive lock for this path.
+
+        // Resolve this Alluxio uri. It should match the original mount id.
+        MountTable.Resolution resolution = mMountTable.resolve(alluxioUri);
+        UnderFileSystem ufs = resolution.getUfs();
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: nit: move this after the if branch

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java [hunk 31]
```diff
@@ -0,0 +1,234 @@
...
+      // Write lock this path, to only enable a single task per path
+      writeLock.lock();
+      ReadWriteLock existingLock = mCurrentPaths.putIfAbsent(alluxioUri.getPath(), rwLock);
+      if (existingLock != null) {
+        // Another thread already locked this path and is processing it. Wait for the other
+        // thread to finish, by locking the existing read lock.
+        writeLock.unlock();
+        writeLock = null;
+        readLock = existingLock.readLock();
+        readLock.lock();
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: it seems like in this case, we will always return `true` ... is that correct?

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java [hunk 32]
```diff
@@ -0,0 +1,234 @@
...
+        return true;
+      }
+      path = path.getParent();
+    }
+    // Reached the root, without finding anything in the cache.
+    return false;
+  }
+
+  /**
+   * Checks the existence of the corresponding ufs path for the given Alluxio path.
```
Comment (jsimsa) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: this method has side-effects (e.g. cache invalidation or insertion), calling it `check` does not communicate this well ... I would call it `process`

File: core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java (hunk 5 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java]: Oops, updated. Thanks.

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java (hunk 32 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: Updated.

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java (hunk 31 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: You are right. I updated it to return false if it the path already exists in the cache.

File: core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java (hunk 30 above)
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java]: Done
